### PR TITLE
Add Batch-Updating Parameter Server Example to CI Tests

### PR DESCRIPTION
**Pull request by H-Huang**
Resolves one item in #46321

This PR sets up DistExamplesTest which will be used as the class to implement future tests for examples. This class is run as part of CI tests. It also creates a dist_examples folder and includes the [batch server example](https://github.com/pytorch/examples/blob/master/distributed/rpc/batch/parameter_server.py) which is slightly modified to allow to be tested.

Run test:
pytest test/distributed/rpc/test_tensorpipe_agent.py -k test_batch_updating_parameter_server -vs
pytest test/distributed/rpc/test_process_group_agent.py -k test_batch_updating_parameter_server -vs

### CODE REVIEW

File: torch/testing/_internal/distributed/rpc/dist_examples/parameter_server.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
+# https://github.com/pytorch/examples/blob/master/distributed/rpc/batch/parameter_server.py
```
Comment (mrshenli) [inline torch/testing/_internal/distributed/rpc/dist_examples/parameter_server.py:1]: You might need to add an empty `__init__.py` so that contents in this folder can be found.

Comment (mrshenli) [reply torch/testing/_internal/distributed/rpc/dist_examples/parameter_server.py:1]: Let's rename this folder just `examples`, as there is already a `distributed` in the path.

File: torch/testing/_internal/distributed/rpc/dist_examples_test.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
+import torch.distributed.rpc as rpc
```
Comment (mrshenli) [inline torch/testing/_internal/distributed/rpc/dist_examples_test.py:1]: Since you've already created a new `examples` folder, we don't need a separate file here. You can merge this test class into `parameter_server.py` and rename it `paramerter_server_test.py`.

See `torch.testing._internal.distributed.rpc.jit.rpc_test` as an example on how tests from sub-folders are imported.

**Review by mrshenli — CHANGES_REQUESTED** [review]

File: torch/testing/_internal/distributed/rpc/dist_examples/parameter_server.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+from datetime import datetime
+
+import torch
+import torch.distributed.rpc as rpc
+import torch.nn as nn
+from torch import optim
+
+batch_size = 20
+image_w = 64
+image_h = 64
```
Comment (mrshenli) [inline torch/testing/_internal/distributed/rpc/dist_examples/parameter_server.py:13]: These two vars are no longer necessary.

File: torch/testing/_internal/distributed/rpc/dist_examples/parameter_server.py [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+
+
+def timed_log(text):
+    print(f"{datetime.now().strftime('%H:%M:%S')} {text}")
+
+
+class BatchUpdateParameterServer(object):
+
+    def __init__(self, batch_update_size=batch_update_size):
+        self.model = nn.Linear(100, num_classes)
```
Comment (mrshenli) [inline torch/testing/_internal/distributed/rpc/dist_examples/parameter_server.py:26]: Since we are replacing module with a simple `Linear` layer, this is no longer called `num_classes`. The first two args in a linear layer are `in_features` and `out_features`. See the doc below. 

https://pytorch.org/docs/stable/generated/torch.nn.Linear.html

File: torch/testing/_internal/distributed/rpc/dist_examples/parameter_server.py [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+
+        return fut
+
+
+class Trainer(object):
+
+    def __init__(self, ps_rref):
+        self.ps_rref = ps_rref
+        self.loss_fn = nn.MSELoss()
+        self.one_hot_indices = torch.LongTensor(batch_size) \
```
Comment (mrshenli) [inline torch/testing/_internal/distributed/rpc/dist_examples/parameter_server.py:67]: Since we are no longer using a ResNet model, we can use a simpler loss function and simpler labels here. This the link below can serve as an example. You don't need to use the same model. The simple `Linear` layer looks good to me. 

https://github.com/pytorch/pytorch/blob/172ed51a17e70aabfbdc096491fde79755de9d08/test/distributed/test_c10d.py#L1928-L1940
https://github.com/pytorch/pytorch/blob/172ed51a17e70aabfbdc096491fde79755de9d08/test/distributed/test_c10d.py#L2073

File: torch/testing/_internal/distributed/rpc/dist_examples/parameter_server.py [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+    timed_log("Start training")
+    ps_rref = rpc.RRef(BatchUpdateParameterServer())
+    futs = []
+    for trainer in trainers:
+        futs.append(
+            rpc.rpc_async(trainer, run_trainer, args=(ps_rref,))
+        )
+
+    torch.futures.wait_all(futs)
+    timed_log("Finish training")
```
Comment (mrshenli) [inline torch/testing/_internal/distributed/rpc/dist_examples/parameter_server.py:108]: Can we log the time spent on training?

File: torch/testing/_internal/distributed/rpc/examples/parameter_server_test.py [hunk 7]
```diff
@@ -0,0 +1,135 @@
+# https://github.com/pytorch/examples/blob/master/distributed/rpc/batch/parameter_server.py
```
Comment (mrshenli) [inline torch/testing/_internal/distributed/rpc/examples/parameter_server_test.py:1]: Let's update this to:

> If you need to modify this file to make this test pass, please also apply same edits accordingly to
> https://github.com/pytorch/examples/blob/master/distributed/rpc/batch/parameter_server.py
> and https://pytorch.org/tutorials/intermediate/rpc_async_execution.html#batch-updating-parameter-server

**Review by mrshenli — COMMENTED** [review]
Code LGTM! Thanks for adding this. Let's wait for tests.

**Review by mrshenli — APPROVED** [review]
LGTM!
